### ISSUE TITLE
autobahn 17.5.1 breaks buildbot unit tests

### ISSUE THREAD

**Issue by tardyp** [crossbario/autobahn-python#832]
```0 % trial buildbot.test.unit.test_mq_wamp.WampMQ.test_forward_data

buildbot.test.unit.test_mq_wamp
  WampMQ
    test_forward_data ...                                               [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/ptardy/dev/bb/buildbot/.venv/lib/python2.7/site-packages/twisted/internet/defer.py", line 1301, in _inlineCallbacks
    result = g.send(result)
  File "/Users/ptardy/dev/bb/buildbot/master/buildbot/test/unit/test_mq_wamp.py", line 135, in test_forward_data
    yield self.mq._produce(('a', 'b'), 'foo')
  File "/Users/ptardy/dev/bb/buildbot/master/buildbot/mq/wamp.py", line 62, in _produce
    return self.master.wamp.publish(self.messageTopic(routingKey), _data, options=options)
  File "/Users/ptardy/dev/bb/buildbot/master/buildbot/test/unit/test_mq_wamp.py", line 70, in publish
    details = EventDetails(1, topic=topic)
exceptions.TypeError: __init__() takes at least 3 arguments (3 given)
```
investigating..

**Comment by tardyp** [comment]
@oberstet looks like it is coming from c2ba9312c83622319168ff3c40f3f2248f937e81

I would prefer this could be done in a backward compatible way. 
from autobahn.wamp.types import EventDetails

is public api, and I would like my test to be able to run on older version of autobahn too.

**Comment by oberstet** [comment]
Ok, there are 2 things here: the concrete issue, and the general issue ("public API").

The constructor has changed: https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/types.py#L482

```python
def __init__(self, subscription, publication, publisher=None, publisher_authid=None, publisher_authrole=None, topic=None, retained=None, enc_algo=None):
```

Likely, your test will work if doing `EventDetails(None, 1, topic=topic)`.

---

Rgd "public API": the _class_ itself is [marked public](https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/types.py#L481), but [not the constructor](https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/types.py#L499).

This marking is intentional: all the properties of `EventDetails` are considered public, but NOT the ctor (there is no user need to construct it).

However we fail in two aspects here:

1. The [docs](https://autobahn.readthedocs.io/en/latest/reference/autobahn.wamp.html#autobahn.wamp.types.EventDetails) do not reflect this properly: the documentation makes no distinction between instance attributes being consider public (as is the case of all of `wamp.types.*`  classes) versus the constructor method itself being public or not
2. There are classes in `wamp.types.*` where the constructor itself must be public, eg `SubscribeOptions` - that is, all classes supposed to be instantiated by the user.

---

To sum up, I take respons. for breaking stuff (sorry:( - and I would like to improve on that;) However, how exactly .. see above .. needs discussions.

What I would like to avoid is a general "every change must work with existing code" policy. This is holding back refactorings, making things more ugly etc.

So I would like to have a crystal clear way of communicating what is public and what is not, and only for the former adhere to a strict policy of not breaking code.

**Comment by tardyp** [comment]
Thanks for detailed answer.
It looks like those @public annotation are fairly recent, and my code is fairly older.

My broken code is an attempt to make a fake WampComponent which does a naive routing itself.

https://github.com/buildbot/buildbot/blob/master/master/buildbot/test/unit/test_mq_wamp.py#L69

so if I cannot construct an EventDetails, the only thing I can see is to actually construct a FakeEventDetails of my own. do you think it is reasonable?

**Comment by oberstet** [comment]
Yeah, this `@public` stuff is quite new, and part of our efforts to make Autobahn more dependable and predictable and "professional" now it is used in production apps people rely on.

In a way, I feel Autobahn move into a new phase, as it is "feature complete" in general (of course there are still many gaps, but it has reached the boundaries of its original scope.

And FWIW, for this phase, I seem to miss some of the language features in strictly/statically typed languages - like enforced privacy etc., and the @public decorator is an attempt to fill this.

> ..make a fake WampComponent

Ah, I see. So it is a unit test within your app (BB) that uses a mock WAMP transport. And since objects of `EventDetails` are provided to user code in the real WAMP transport, a fake transport needs to construct it.

I think this busts my naive thinking: "there is no user need to construct it" =)

IOW: in this particular instance, I'd say we should make the constructor public too (`EventDetails.__init__` should be marked `@public`) and be bound with the responsibility that comes with that.

**Comment by tardyp** [comment]
As we could not merge anything in our mainline, I chose to use the simple fake object.
https://github.com/buildbot/buildbot/pull/3164

I like the @public decorator. In buildbot, we have the policy of keeping compat for everything that is documented. We have a separated, non-autogenerated documentation for that.

This is often painful to keep backward compatibility, but generally not that much impossible.
Python being a dynamic language also helps on that :)

What we usually do is to only add keyword arguments and we don't modify positional arguments.
In the case of EventDetails, I think this would have been pretty simple to do it, at the price of a bit more verbosity for the caller.
We don't use positional as way to mark arguments mandatory, but rather make an assertion that its not configured with the default value.

m2c

**Comment by tardyp** [comment]
sorry ``@public`` to mistakinly include you in this discussion. Not that a good idea for a github nick :)
